src/internal_events/codecs.rs

```diff
@@ -0,0 +1,92 @@
+use metrics::counter;
+use vector_core::internal_event::InternalEvent;
+
+use super::prelude::{error_stage, error_type};
+
+#[derive(Debug)]
+pub struct DecoderFramingFailed<'a> {
+    pub error: &'a codecs::decoding::BoxedFramingError,
+}
+
+impl<'a> InternalEvent for DecoderFramingFailed<'a> {
+    fn emit(self) {
+        warn!(message = "Failed framing bytes.", error = %self.error, internal_log_rate_secs = 10);
+        counter!("decoder_framing_errors_total", 1);
+    }
+}
+
+#[derive(Debug)]
+pub struct DecoderDeserializeFailed<'a> {
+    pub error: &'a crate::Error,
+}
+
+impl<'a> InternalEvent for DecoderDeserializeFailed<'a> {
+    fn emit(self) {
+        warn!(message = "Failed deserializing frame.", error = %self.error, internal_log_rate_secs = 10);
+        counter!("decoder_deserialize_errors_total", 1);
+    }
+}
+
+#[derive(Debug)]
+pub struct EncoderFramingError<'a> {
+    pub error: &'a codecs::encoding::BoxedFramingError,
+}
+
+impl<'a> InternalEvent for EncoderFramingError<'a> {
+    fn emit(self) {
+        warn!(message = "Failed framing bytes.", error = %self.error, internal_log_rate_secs = 10);
+        error!(
+            message = "Events dropped.",
+            count = 1,
+            error = %self.error,
+            error_type = error_type::ENCODER_FAILED,
+            stage = error_stage::SENDING,
+            intentional = "false",
+            reason = "Failed framing bytes.",
+        );
+        counter!("encoder_framing_errors_total", 1);
+        counter!(
+            "component_errors_total", 1,
+            "error_type" => error_type::ENCODER_FAILED,
+            "stage" => error_stage::SENDING,
+        );
+        counter!(
+            "component_discarded_events_total", 1,
+            "error_type" => error_type::ENCODER_FAILED,
+            "stage" => error_stage::SENDING,
+            "intentional" => "false",
+        );
```

Comment:
I might be missing something but shouldn't this metric include `count` as described here:

https://github.com/vectordotdev/vector/blob/master/docs/specs/instrumentation.md#eventsdropped

?

Reply:
Based on what's written the counter should indeed bear the `count` tag but it seems kind of redundant and not super useful (and existing event that seems to comply with the instrumentation specs don't do it https://github.com/vectordotdev/vector/blob/master/src/internal_events/common.rs#L87-L99).

Reply:
I see, I'll drop this per https://github.com/vectordotdev/vector/pull/14091 
Thanks for clearing it up :+1:

Reply:
Yeah, I had raised this earlier as well - a `count` tag is superfluous on a `counter` metric.